### PR TITLE
Add storage-class support

### LINKED ISSUE #1859: s3: [feature] Please add support for using different storage classes in S3
Commands like `mc cp` and `mc mirror` should take a `--storage-class` parameter which would specify the storage class for newly-created objects. Right now, everything (reasonably) defaults to `STANDARD`.

### PR DESCRIPTION
**Pull request by nitisht**
This PR adds storage-class support for mc cp and mirror command. Storage class
values are passed via the `-storage-class` or `--sc` flag.

Fixes #1859

### CODE REVIEW

File: cmd/cp-main.go [hunk 1]
```diff
@@ -317,6 +321,15 @@ func doCopySession(session *sessionV8) error {
 			// Save totalSize.
 			cpURLs.TotalSize = session.Header.TotalBytes
 
+			// Save storage class metadata if present
+			if cpURLs.TargetContent.Metadata == nil {
+				cpURLs.TargetContent.Metadata = make(map[string]string)
+			}
+
+			if _, ok := session.Header.CommandStringFlags["storage-class"]; ok {
```
Comment (kannappanr) [inline cmd/cp-main.go]: The check for the existence of `cpURLs.TargetContent.Metadata` can be moved inside the `if` block, that way we do not create an empty map in certain cases.

File: docs/minio-client-complete-guide.md [hunk 2]
```diff
@@ -352,8 +352,9 @@ USAGE:
    mc cp [FLAGS] SOURCE [SOURCE...] TARGET
    
 FLAGS:
-  --help, -h                       Show help.
-  --recursive, -r		   Copy recursively.
+  --recursive, -r                    Copy recursively.
+  --storage-class value, --sc value  Set storage class for object.
```
Comment (kannappanr) [inline docs/minio-client-complete-guide.md]: Should `--storage-class value, --sc value  Set storage class for object.` be `--storage-class value, -sc value  Set storage class for object.` ?

File: docs/minio-client-complete-guide.md [hunk 3]
```diff
@@ -362,6 +363,14 @@ FLAGS:
 mc cp myobject.txt play/mybucket
 myobject.txt:    14 B / 14 B  ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓  100.00 % 41 B/s 0
 ```
+
+*Example: Copy a text file to to an object storage and assign storage-class `REDUCED_REDUNDANCY` to the uploaded object.*
+
+```sh
+mc cp -storage-class REDUCED_REDUNDANCY myobject.txt play/mybucket
```
Comment (kannappanr) [inline docs/minio-client-complete-guide.md]: Should this be changed to `mc cp -storage-class REDUCED_REDUNDANCY myobject.txt play/mybucket` to `mc cp --storage-class REDUCED_REDUNDANCY myobject.txt play/mybucket` ?

**Comment by nitisht** [comment]
Updated the PR @kannappanr PTAL

File: docs/minio-client-complete-guide.md [hunk 4]
```diff
@@ -362,6 +363,14 @@ FLAGS:
 mc cp myobject.txt play/mybucket
 myobject.txt:    14 B / 14 B  ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓  100.00 % 41 B/s 0
 ```
+
+*Example: Copy a text file to to an object storage and assign storage-class `REDUCED_REDUNDANCY` to the uploaded object.*
```
Comment (poornas) [inline docs/minio-client-complete-guide.md]: s/to to an object storage/to an object storage/

File: functional-tests.sh [hunk 5]
```diff
@@ -530,6 +530,18 @@ function test_config_host_add_error()
...
+function test_put_object_with_storage_class()
+{
+    show "${FUNCNAME[0]}"
+
+    start_time=$(get_time)
+    object_name="mc-test-object-$RANDOM"
+    assert_success "$start_time" "${FUNCNAME[0]}" mc_cmd cp --storage-class REDUCED_REDUNDANCY "${FILE_1_MB}" "${SERVER_ALIAS}/${BUCKET_NAME}/${object_name}"
+    assert_success "$start_time" "${FUNCNAME[0]}" mc_cmd rm "${SERVER_ALIAS}/${BUCKET_NAME}/${object_name}"
+
+    log_success "$start_time" "${FUNCNAME[0]}"
```
Comment (poornas) [inline functional-tests.sh]: can you add a negative test with invalid storage class as well?

Comment (nitisht) [reply functional-tests.sh]: done

**Comment by nitisht** [comment]
> The referenced issue asks for storage class support on both cp and mirror operations. don't see --storage-class added to mc mirror here.

Added support for mc mirror as well. PTAL

File: test/RED [hunk 6]
```diff
@@ -0,0 +1,188 @@
+# Minio Client Quickstart Guide
```
Comment (harshavardhana) [inline test/RED]: What is this file for?

Comment (nitisht) [reply test/RED]: oops wrongly committed, removed

File: docs/minio-client-complete-guide.md [hunk 7]
```diff
@@ -495,11 +504,12 @@ USAGE:
...
-  --force			   Force overwrite of an existing target(s).
-  --fake			   Perform a fake mirror operation.
-  --watch, -w                      Watch and mirror for changes.
-  --remove			   Remove extraneous file(s) on target.
+  --help, -h                         Show help.
+  --force			                       Force overwrite of an existing target(s).
+  --fake			                       Perform a fake mirror operation.
+  --watch, -w                        Watch and mirror for changes.
+  --remove			                     Remove extraneous file(s) on target.
+  --storage-class value, --sc value  Set storage class for object.
```
Comment (poornas) [inline docs/minio-client-complete-guide.md]: indentation issue here.

**Comment by kannappanr** [comment]
Except for the indentation issue mentioned above, everything looks good to me. Tested it against Play.

File: docs/minio-client-complete-guide.md (hunk 7 above)
Comment (nitisht) [reply docs/minio-client-complete-guide.md]: fixed

**Comment by poornas** [comment]
@nitisht, when I enter storage class in lowercase, the mc cp command fails - the input should be upper cased instead of erroring out.
```
➜  mc git:(11fe81e4) mc cp ~/Downloads/smallfile myminio/testbucket/sc1 --storage-class standard
mc: <ERROR> Failed to copy `/home/kris/Downloads/smallfile`. Invalid storage class.
mc: <ERROR> Session safely terminated. To resume session `mc session resume SvDATtTm`
➜  mc git:(11fe81e4) mc cp ~/Downloads/smallfile myminio/testbucket/sc1 --storage-class STANDARD
/home/kris/Downloads/smallfile:   132 B / 132 B ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 100.00% 18.38 KB/s 0s
```

**Comment by harshavardhana** [comment]
> ➜  mc git:(11fe81e4) mc cp ~/Downloads/smallfile myminio/testbucket/sc1 --storage-class standard
> mc: <ERROR> Failed to copy `/home/kris/Downloads/smallfile`. Invalid storage class.
> mc: <ERROR> Session safely terminated. To resume session `mc session resume SvDATtTm`
> ➜  mc git:(11fe81e4) mc cp ~/Downloads/smallfile myminio/testbucket/sc1 --storage-class STANDARD

We can simply support both lower and upper, just that we send upper to server. @nitisht ?

**Comment by nitisht** [comment]
> We can simply support both lower and upper, just that we send upper to server. @nitisht ?

Agreed @harshavardhana , updated the PR accordingly. PTAL @poornas

**Pull request merged**